File: templates/web-app/backend/app/services/auth_service.py
"""Auth business logic: register, login, refresh (RTR), logout.

Refresh Token Rotation (RTR) + reuse detection (AGENTS.md §4.1 / ADR-002):
- Every refresh issues a NEW token pair and revokes the old one.
- If an already-revoked token is presented (replay), ALL descendant sessions
  for that user are revoked and a security incident is audited.
"""
from datetime import datetime, timezone

from redis.asyncio import Redis
from sqlalchemy import select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.core import security
from app.core.config import settings
from app.models import AuthToken, RefreshToken, User
from app.repositories import create_user, get_user_by_email, get_user_by_id, revoke_all_refresh_tokens
from app.services.audit_service import write_audit


class AuthError(Exception):
    def __init__(self, code: str, message: str):
        self.code = code
        self.message = message


PURPOSE_PASSWORD_RESET = "password_reset"
PURPOSE_EMAIL_VERIFICATION = "email_verification"


async def register(db: AsyncSession, email: str, password: str) -> User:
    if await get_user_by_email(db, email):
        raise AuthError("VALIDATION_ERROR", "Email already registered")
    user = await create_user(db, email, security.hash_password(password))
    await write_audit(db, actor_user_id=user.id, action="user.register", resource=f"user:{user.id}")
    return user


async def _issue_token_pair(db: AsyncSession, user: User, parent_id: str | None = None) -> tuple[str, str]:
    access = security.create_access_token(subject=user.id, role=user.role)
    refresh = security.new_refresh_token()
    db.add(RefreshToken(
        user_id=user.id,
        token_hash=security.hash_refresh_token(refresh),
        parent_id=parent_id,
        expires_at=security.refresh_token_expiry(),
    ))
    await db.commit()
    return access, refresh


async def authenticate(db: AsyncSession, email: str, password: str) -> User:
    """Check credentials and account state.

    The shared precondition of every login path, so MFA and plain login cannot
    drift on what "valid credentials" means.
    """
    user = await get_user_by_email(db, email)
    if not user or not security.verify_password(password, user.hashed_password):
        raise AuthError("UNAUTHORIZED", "Invalid credentials")
    if user.status != "active":
        raise AuthError("FORBIDDEN", "Account is not active")
    return user


async def login(db: AsyncSession, email: str, password: str) -> tuple[str, str, int]:
    user = await authenticate(db, email, password)
    return await issue_session(db, user)


async def issue_session(db: AsyncSession, user: User) -> tuple[str, str, int]:
    """Full login for a user who has passed every factor."""
    access, refresh = await _issue_token_pair(db, user)
    await write_audit(db, actor_user_id=user.id, action="user.login", resource=f"user:{user.id}")
    return access, refresh, settings.ACCESS_TOKEN_EXPIRE_MINUTES * 60


async def issue_mfa_session(db: AsyncSession, user: User) -> str:
    """Half-authenticated session: a refresh token exists, but no access token
    is issued until the second factor is proven (ADR-008)."""
    _, refresh = await _issue_token_pair(db, user)
    await write_audit(
        db,
        actor_user_id=user.id,
        action="user.login_mfa_challenge",
        resource=f"user:{user.id}",
    )
    return refresh


async def complete_mfa_login(
    db: AsyncSession, redis: Redis, presented_token: str, code: str
) -> tuple[str, str, int]:
    """Finish an MFA login: validate the pending refresh token and the TOTP
    code, then rotate to a fully authenticated session."""
    from app.services import mfa_service

    token_hash = security.hash_refresh_token(presented_token)
    res = await db.execute(select(RefreshToken).where(RefreshToken.token_hash == token_hash))
    stored = res.scalar_one_or_none()
    if stored is None:
        raise AuthError("UNAUTHORIZED", "Invalid refresh token")

    now = datetime.now(timezone.utc)
    expires_at = stored.expires_at
    if expires_at.tzinfo is None:
        expires_at = expires_at.replace(tzinfo=timezone.utc)
    if stored.revoked_at is not None or expires_at < now:
        raise AuthError("UNAUTHORIZED", "Invalid refresh token")

    res_user = await db.execute(select(User).where(User.id == stored.user_id))
    user = res_user.scalar_one_or_none()
    if not user or not user.mfa_enabled:
        raise AuthError("FORBIDDEN", "MFA is not required for this account")

    method = await mfa_service.verify_code(db, redis, user, code)
    if method is None:
        await write_audit(
            db,
            actor_user_id=user.id,
            action="user.mfa.login_failed",
            resource=f"user:{user.id}",
            detail="Failed second factor on MFA login",
        )
        await db.commit()
        raise AuthError("UNAUTHORIZED", "Invalid TOTP code")

    stored.revoked_at = now
    access, new_refresh = await _issue_token_pair(db, user, parent_id=stored.id)
    await write_audit(
        db,
        actor_user_id=user.id,
        action="user.login",
        resource=f"user:{user.id}",
        detail=f"Login completed via MFA ({method})",
    )
    return access, new_refresh, settings.ACCESS_TOKEN_EXPIRE_MINUTES * 60


async def refresh(db: AsyncSession, presented_token: str) -> tuple[str, str, int]:
    token_hash = security.hash_refresh_token(presented_token)
    res = await db.execute(select(RefreshToken).where(RefreshToken.token_hash == token_hash))
    stored = res.scalar_one_or_none()
    now = datetime.now(timezone.utc)

    if stored is None:
        raise AuthError("UNAUTHORIZED", "Invalid refresh token")

    if stored.revoked_at is not None:
        # REPLAY: an old token was reused -> revoke ALL sessions for the user.
        await db.execute(
            update(RefreshToken)
            .where(RefreshToken.user_id == stored.user_id, RefreshToken.revoked_at.is_(None))
            .values(revoked_at=now)
        )
        await write_audit(
            db, actor_user_id=stored.user_id,
            action="security.refresh_token_reuse_detected",
            resource=f"user:{stored.user_id}",
            detail="Revoked all sessions after refresh-token replay",
        )
        await db.commit()
        raise AuthError("UNAUTHORIZED", "Session revoked due to token reuse")

    if stored.expires_at < now:
        raise AuthError("UNAUTHORIZED", "Refresh token expired")

    # Rotate: revoke old, issue new pair chained to it.
    stored.revoked_at = now
    res_user = await db.execute(select(User).where(User.id == stored.user_id))
    user = res_user.scalar_one()
    access, new_refresh = await _issue_token_pair(db, user, parent_id=stored.id)
    return access, new_refresh, settings.ACCESS_TOKEN_EXPIRE_MINUTES * 60


async def logout(db: AsyncSession, presented_token: str) -> None:
    token_hash = security.hash_refresh_token(presented_token)
    now = datetime.now(timezone.utc)
    await db.execute(
        update(RefreshToken)
        .where(RefreshToken.token_hash == token_hash, RefreshToken.revoked_at.is_(None))
        .values(revoked_at=now)
    )
    await db.commit()


async def _claim_auth_token(db: AsyncSession, raw_token: str, purpose: str) -> AuthToken:
    """Look up an emailed token and validate it. Consumes nothing on failure.

    The three failure modes are deliberately indistinguishable to the caller —
    "no such token", "already used" and "expired" all answer the same way, so a
    guess cannot learn which links might still be live.
    """
    res = await db.execute(
        select(AuthToken).where(
            AuthToken.token_hash == security.hash_token(raw_token),
            AuthToken.purpose == purpose,
        )
    )
    record = res.scalar_one_or_none()
    if record is None:
        raise AuthError("UNAUTHORIZED", "Invalid link")

    if record.consumed_at is not None:
        raise AuthError("UNAUTHORIZED", "Invalid link")

    # SQLite returns naive datetimes; Postgres returns tz-aware ones. Normalise
    # so the comparison is well-defined on either driver (same fix as ADR-006).
    expires_at = record.expires_at
    if expires_at.tzinfo is None:
        expires_at = expires_at.replace(tzinfo=timezone.utc)
    if expires_at < datetime.now(timezone.utc):
        raise AuthError("UNAUTHORIZED", "Invalid link")

    return record


async def request_password_reset(db: AsyncSession, email: str) -> str | None:
    """Create a single-use reset token. Returns the raw token, or None if no
    account matches.

    None is deliberately not an error: this endpoint is unauthenticated, and
    distinguishing "sent" from "no such user" would turn it into an
    account-enumeration oracle. The caller answers 200 either way.
    """
    user = await get_user_by_email(db, email)
    if not user:
        return None

    token = security.new_opaque_token()
    db.add(AuthToken(
        user_id=user.id,
        token_hash=security.hash_token(token),
        purpose=PURPOSE_PASSWORD_RESET,
        expires_at=security.password_reset_token_expiry(),
    ))
    await write_audit(
        db,
        actor_user_id=user.id,
        action="user.password_reset_requested",
        resource=f"auth:reset:{security.email_fingerprint(email)}",
    )
    await db.commit()
    return token


async def confirm_password_reset(db: AsyncSession, raw_token: str, new_password: str) -> User:
    """Redeem a reset token and set the new password.

    The password change and the session revocation run in one transaction: a
    browser tab hijacked before the reset must not keep working on the old
    credentials (same blast-radius rule as the JML kill-switch, ADR-005).
    """
    record = await _claim_auth_token(db, raw_token, PURPOSE_PASSWORD_RESET)

    user = await get_user_by_id(db, record.user_id)
    if not user:
        raise AuthError("NOT_FOUND", "Account no longer exists")
    if user.status != "active":
        raise AuthError("FORBIDDEN", "Account is not active")

    record.consumed_at = datetime.now(timezone.utc)
    user.hashed_password = security.hash_password(new_password)
    sessions_revoked = await revoke_all_refresh_tokens(db, user.id)

    await write_audit(
        db,
        actor_user_id=user.id,
        action="user.password_reset",
        resource=f"user:{user.id}",
        detail=f"Password changed via reset token; {sessions_revoked} session(s) revoked",
    )
    await db.commit()
    await db.refresh(user)
    return user


async def request_email_verification(db: AsyncSession, user_id: str) -> str:
    """Create a single-use verification token for the caller's own address."""
    user = await get_user_by_id(db, user_id)
    if not user:
        raise AuthError("NOT_FOUND", "Account not found")
    if user.email_verified_at is not None:
        raise AuthError("VALIDATION_ERROR", "Email is already verified")

    token = security.new_opaque_token()
    db.add(AuthToken(
        user_id=user.id,
        token_hash=security.hash_token(token),
        purpose=PURPOSE_EMAIL_VERIFICATION,
        expires_at=security.email_verification_token_expiry(),
    ))
    await write_audit(
        db,
        actor_user_id=user.id,
        action="user.email_verification_requested",
        resource=f"user:{user.id}",
    )
    await db.commit()
    return token


async def confirm_email_verification(db: AsyncSession, raw_token: str) -> User:
    """Redeem a verification token and mark the address verified."""
    record = await _claim_auth_token(db, raw_token, PURPOSE_EMAIL_VERIFICATION)

    user = await get_user_by_id(db, record.user_id)
    if not user:
        raise AuthError("NOT_FOUND", "Account no longer exists")

    record.consumed_at = datetime.now(timezone.utc)
    user.email_verified_at = datetime.now(timezone.utc)

    await write_audit(
        db,
        actor_user_id=user.id,
        action="user.email_verified",
        resource=f"user:{user.id}",
    )
    await db.commit()
    await db.refresh(user)
    return user

File: templates/web-app-express/backend/src/modules/auth/auth.service.ts
import crypto from "crypto";
import { query } from "../../core/db.js";
import {
  hashPassword,
  verifyPassword,
  createAccessToken,
  createRefreshToken,
  verifyRefreshToken,
  hashToken,
} from "../../core/crypto.js";
import { logAuditEvent } from "../../core/audit.js";

export class AuthService {
  static async register(email: string, password: string, ip: string) {
    const existing = await query("SELECT id FROM users WHERE email = $1", [email]);
    if (existing.rows.length > 0) {
      throw { statusCode: 409, code: "EMAIL_EXISTS", message: "Email is already registered" };
    }

    const hashedPassword = await hashPassword(password);
    const result = await query(
      `INSERT INTO users (email, password_hash, role, status)
       VALUES ($1, $2, 'user', 'active')
       RETURNING id, email, role, status, created_at`,
      [email, hashedPassword]
    );

    const newUser = result.rows[0];
    await logAuditEvent({
      userId: newUser.id,
      actorRole: "user",
      action: "USER_REGISTERED",
      resourceType: "user",
      resourceId: newUser.id,
      ipAddress: ip,
      status: "SUCCESS",
    });

    return newUser;
  }

  static async login(email: string, password: string, ip: string) {
    const userRes = await query(
      `SELECT id, email, password_hash, role, status, failed_login_attempts, locked_until 
       FROM users WHERE email = $1`,
      [email]
    );

    if (userRes.rows.length === 0) {
      throw { statusCode: 401, code: "INVALID_CREDENTIALS", message: "Invalid email or password" };
    }

    const user = userRes.rows[0];

    // 1. Account Lockout check (ADR-004)
    if (user.locked_until && new Date(user.locked_until) > new Date()) {
      const remainingSeconds = Math.ceil(
        (new Date(user.locked_until).getTime() - Date.now()) / 1000
      );
      throw {
        statusCode: 423,
        code: "ACCOUNT_LOCKED",
        message: `Account is temporarily locked. Try again in ${remainingSeconds} seconds.`,
        details: { retryAfter: remainingSeconds },
      };
    }

    // 2. JML Status check (ADR-005)
    if (user.status !== "active") {
      throw {
        statusCode: 403,
        code: "ACCOUNT_DISABLED",
        message: "Account is suspended or terminated",
      };
    }

    // 3. Password Verification
    const isMatch = await verifyPassword(password, user.password_hash);
    if (!isMatch) {
      const attempts = user.failed_login_attempts + 1;
      let lockedUntil: Date | null = null;

      if (attempts >= 5) {
        lockedUntil = new Date(Date.now() + 15 * 60 * 1000); // 15 min lock
        await query(
          "UPDATE users SET failed_login_attempts = $1, locked_until = $2 WHERE id = $3",
          [attempts, lockedUntil, user.id]
        );
        await logAuditEvent({
          userId: user.id,
          actorRole: user.role,
          action: "ACCOUNT_LOCKED_FAILED_ATTEMPTS",
          resourceType: "user",
          resourceId: user.id,
          ipAddress: ip,
          status: "FAILURE",
          details: { attempts },
        });
        throw {
          statusCode: 423,
          code: "ACCOUNT_LOCKED",
          message: "Account locked for 15 minutes due to 5 consecutive failed login attempts.",
          details: { retryAfter: 900 },
        };
      } else {
        await query("UPDATE users SET failed_login_attempts = $1 WHERE id = $2", [
          attempts,
          user.id,
        ]);
        await logAuditEvent({
          userId: user.id,
          actorRole: user.role,
          action: "LOGIN_FAILED",
          resourceType: "user",
          resourceId: user.id,
          ipAddress: ip,
          status: "FAILURE",
          details: { attempts },
        });
        throw {
          statusCode: 401,
          code: "INVALID_CREDENTIALS",
          message: "Invalid email or password",
        };
      }
    }

    // 4. Success — Reset lockout counters
    await query("UPDATE users SET failed_login_attempts = 0, locked_until = NULL WHERE id = $1", [
      user.id,
    ]);

    // 5. Issue Tokens & Setup RTR Session
    const familyId = crypto.randomUUID();
    const accessToken = createAccessToken({ userId: user.id, role: user.role });
    const refreshToken = createRefreshToken({
      userId: user.id,
      familyId,
      jti: crypto.randomUUID(),
    });
    const tokenHash = hashToken(refreshToken);
    const expiresAt = new Date(Date.now() + 7 * 24 * 60 * 60 * 1000);

    await query(
      `INSERT INTO refresh_tokens (user_id, token_hash, family_id, expires_at)
       VALUES ($1, $2, $3, $4)`,
      [user.id, tokenHash, familyId, expiresAt]
    );

    await logAuditEvent({
      userId: user.id,
      actorRole: user.role,
      action: "LOGIN_SUCCESS",
      resourceType: "user",
      resourceId: user.id,
      ipAddress: ip,
      status: "SUCCESS",
    });

    return {
      accessToken,
      refreshToken,
      user: {
        id: user.id,
        email: user.email,
        role: user.role,
        status: user.status,
      },
    };
  }

  static async refresh(oldRefreshToken: string, ip: string) {
    let decoded: any;
    try {
      decoded = verifyRefreshToken(oldRefreshToken);
    } catch (err) {
      throw { statusCode: 401, code: "INVALID_TOKEN", message: "Invalid or expired refresh token" };
    }

    const tokenHash = hashToken(oldRefreshToken);
    const tokenRes = await query(
      "SELECT id, user_id, family_id, is_revoked FROM refresh_tokens WHERE token_hash = $1",
      [tokenHash]
    );

    if (tokenRes.rows.length === 0) {
      throw { statusCode: 401, code: "INVALID_TOKEN", message: "Token not found" };
    }

    const record = tokenRes.rows[0];

    // 🚨 RTR Replay Detection: An already-revoked token was presented!
    if (record.is_revoked) {
      // Revoke all tokens in this family immediately!
      await query("UPDATE refresh_tokens SET is_revoked = TRUE WHERE family_id = $1", [
        record.family_id,
      ]);
      await logAuditEvent({
        userId: record.user_id,
        actorRole: "user",
        action: "TOKEN_REPLAY_DETECTED_ALL_REVOKED",
        resourceType: "refresh_token",
        resourceId: record.family_id,
        ipAddress: ip,
        status: "DENIED",
      });
      throw {
        statusCode: 401,
        code: "TOKEN_REPLAY_DETECTED",
        message: "Security violation: Compromised token reuse detected. All active sessions revoked.",
      };
    }

    // Revoke the old token (Single-use rotation)
    await query("UPDATE refresh_tokens SET is_revoked = TRUE WHERE id = $1", [record.id]);

    // Check user status
    const userRes = await query("SELECT id, role, status FROM users WHERE id = $1", [
      record.user_id,
    ]);
    if (userRes.rows.length === 0 || userRes.rows[0].status !== "active") {
      throw { statusCode: 403, code: "ACCOUNT_DISABLED", message: "Account is no longer active" };
    }

    const user = userRes.rows[0];

    // Issue new rotated pair keeping the same family
    const accessToken = createAccessToken({ userId: user.id, role: user.role });
    const newRefreshToken = createRefreshToken({
      userId: user.id,
      familyId: record.family_id,
      jti: crypto.randomUUID(),
    });
    const newTokenHash = hashToken(newRefreshToken);
    const expiresAt = new Date(Date.now() + 7 * 24 * 60 * 60 * 1000);

    await query(
      `INSERT INTO refresh_tokens (user_id, token_hash, family_id, expires_at)
       VALUES ($1, $2, $3, $4)`,
      [user.id, newTokenHash, record.family_id, expiresAt]
    );

    await logAuditEvent({
      userId: user.id,
      actorRole: user.role,
      action: "TOKEN_ROTATION_SUCCESS",
      resourceType: "refresh_token",
      ipAddress: ip,
      status: "SUCCESS",
    });

    return { accessToken, newRefreshToken };
  }

  static async logout(refreshToken?: string) {
    if (refreshToken) {
      const tokenHash = hashToken(refreshToken);
      await query("UPDATE refresh_tokens SET is_revoked = TRUE WHERE token_hash = $1", [tokenHash]);
    }
  }
}
